### PR TITLE
Report md5sum workflow duration to BigQuery

### PR DESCRIPTION
**Pull request by jessebrennan**

### CODE REVIEW

File: test/test_basic_submission.py [hunk 1]
```diff
@@ -109,36 +110,47 @@ def test_dockstore_import_in_terra(self):
...
+                                           f'Expected 4 minutes, but took longer than '
+                                           f'{float(start - now) / 60.0} minutes.')
+
+            with self.subTest('Dockstore Workflow Run Completed Successfully'):
+                if response['workflows'][0]['status'] != "Succeeded":
+                    raise RuntimeError(f'The md5sum workflow did not succeed:\n{json.dumps(response, indent=4)}')
+        finally:
+            table = f'platform-dev-178517.bdc.terra_md5_latency_{STAGE}'
+            client = Client()
+            client.add_row(table, time.time() - start)
```
Comment (DailyDreaming) [inline test/test_basic_submission.py:153]: This should probably have it's own function to add the row with a retry decorator on it.

There should probably be an extra try except here too when calling that function, since if it does not post the row after multiple retries (because BigQuery is down for whatever reason), but the test passed, we should still report the test success/failure and not fail on the latency post.

File: test/test_basic_submission.py [hunk 2]
```diff
@@ -109,36 +110,47 @@ def test_dockstore_import_in_terra(self):
...
+                    else:
+                        print(json.dumps(response, indent=4))
+                        raise RuntimeError('The md5sum workflow run timed out.  '
+                                           f'Expected 4 minutes, but took longer than '
+                                           f'{float(start - now) / 60.0} minutes.')
+
+            with self.subTest('Dockstore Workflow Run Completed Successfully'):
+                if response['workflows'][0]['status'] != "Succeeded":
+                    raise RuntimeError(f'The md5sum workflow did not succeed:\n{json.dumps(response, indent=4)}')
+        finally:
```
Comment (DailyDreaming) [inline test/test_basic_submission.py:150]: We should only post the time if we get a successful workflow submission.  Otherwise it's going to be posting very quick latency times whenever the test fails, which it shouldn't do.

**Review by DailyDreaming — CHANGES_REQUESTED** [review]

File: test/test_basic_submission.py (hunk 1 above)
Comment (jessebrennan) [reply test/test_basic_submission.py:153]: > This should probably have its own function to add the row with a retry decorator on it.

:+1:

> There should probably be an extra try except here too when calling that function

I considered that, but then we wouldn't know if the latency failed to report.

File: test/test_basic_submission.py (hunk 2 above)
Comment (jessebrennan) [reply test/test_basic_submission.py:150]: Hmm interest point. Quick failures like that could be confusing in the graph. If a workflow queues for 15min then fails, wouldn't we want to capture that information though?

File: test/test_basic_submission.py (hunk 1 above)
Comment (jessebrennan) [reply test/test_basic_submission.py:153]: I'll add a try except.

File: test/test_basic_submission.py (hunk 2 above)
Comment (DailyDreaming) [reply test/test_basic_submission.py]: Yeah, we should capture failures and successes, but only if the initial workflow request succeeds.  We want to measure the time it's queued and running.  If it's never queued, we should discard that though, since the workflow never ran and that's not a measure of latency.

File: test/test_basic_submission.py [hunk 3]
```diff
@@ -120,25 +121,42 @@ def test_drs_workflow_in_terra(self):
...
                 raise RuntimeError(f'The md5sum workflow did not succeed:\n{json.dumps(response, indent=4)}')
+            else:
+                self.log_runtime(table, time.time() - start)
+
+    def log_runtime(self, table, start):
+        try:
+            Client().add_row(table, time.time() - start)
+        except Exception:
+            # We don't want failed logging to fail the whole test
+            logger.warning('Failed to log run tim to BigQuery')
```
Comment (DailyDreaming) [inline test/test_basic_submission.py:158]: `tim`

File: test/test_basic_submission.py [hunk 4]
```diff
@@ -120,25 +121,42 @@ def test_drs_workflow_in_terra(self):
...
+                else:
+                    print(json.dumps(response, indent=4))
+                    self.log_runtime(table, time.time() - start)
+                    raise RuntimeError('The md5sum workflow run timed out.  '
+                                       f'Expected 4 minutes, but took longer than '
+                                       f'{float(start - now) / 60.0} minutes.')
 
         with self.subTest('Dockstore Workflow Run Completed Successfully'):
             if response['workflows'][0]['status'] != "Succeeded":
                 raise RuntimeError(f'The md5sum workflow did not succeed:\n{json.dumps(response, indent=4)}')
```
Comment (DailyDreaming) [inline test/test_basic_submission.py:149]: We should log the runtime here too.

File: test/test_basic_submission.py [hunk 5]
```diff
@@ -120,25 +121,42 @@ def test_drs_workflow_in_terra(self):
...
+        start = time.time()
+        deadline = start + int(os.environ.get('MD5SUM_TEST_TIMEOUT', 60 * 60))
+        table = f'platform-dev-178517.bdc.terra_md5_latency_{STAGE}'
+        while True:
             response = check_workflow_status(submission_id=submission_id)
-            time.sleep(20)
-            timeout -= 20
             status = response['status']
             if response['workflows'][0]['status'] == "Failed":
                 raise RuntimeError(f'The md5sum workflow did not succeed:\n{json.dumps(response, indent=4)}')
```
Comment (DailyDreaming) [inline test/test_basic_submission.py:131]: We should log the runtime here too.

File: test/test_basic_submission.py [hunk 6]
```diff
@@ -120,25 +121,42 @@ def test_drs_workflow_in_terra(self):
 
         # md5sum should run for about 4 minutes, but may take far longer(?); give a generous timeout
         # also configurable manually via MD5SUM_TEST_TIMEOUT if held in a pending state
-        timeout = twenty_minutes = int(os.environ.get('MD5SUM_TEST_TIMEOUT', 60 * 60))
-        while status != 'Done':
+        start = time.time()
+        deadline = start + int(os.environ.get('MD5SUM_TEST_TIMEOUT', 60 * 60))
+        table = f'platform-dev-178517.bdc.terra_md5_latency_{STAGE}'
+        while True:
```
Comment (DailyDreaming) [inline test/test_basic_submission.py:127]: Why was this changed to `while True`?  Faster exit?

**Review by DailyDreaming — CHANGES_REQUESTED** [review]

**Review by DailyDreaming — APPROVED** [review]
This looks like it should work.
